dashboard/modules/job/job_head.py

```diff
@@ -204,6 +212,80 @@ async def list_jobs(self, req: Request) -> Response:
             content_type="application/json",
         )
 
+    @routes.get("/api/all_jobs/")
+    @optional_utils.init_ray_and_catch_exceptions(connect_to_serve=False)
+    async def list_jobs(self, req: Request) -> Response:
+        driver_jobs, submission_job_drivers = await self._get_driver_jobs()
+        submission_jobs = self._job_manager.list_jobs()
+        for job_id, job in submission_jobs.items():
+            job.driver = submission_job_drivers.get(job_id)
+        return Response(
+            text=json.dumps(
+                [
+                    *[
+                        {
+                            **dataclasses.asdict(job_info),
+                            "id": job_id,
+                            "type": "submission",
+                        }
+                        for job_id, job_info in submission_jobs.items()
+                    ],
+                    *[
+                        {**dataclasses.asdict(job_info), "id": job_id, "type": "driver"}
+                        for job_id, job_info in driver_jobs.items()
+                    ],
+                ]
+            ),
+            content_type="application/json",
+        )
+
+    async def _get_driver_jobs(self):
+        """
+        Returns a tuple of dictionaries related to drivers.
+
+        The first dictionary contains all driver jobs and is keyed by the job's id.
+        The second dictionary contains drivers that belong to submission jobs.
+        It's keyed by the submission job's id. Only the last driver of a submission job is returned.
+        """
+        request = gcs_service_pb2.GetAllJobInfoRequest()
+        reply = await self._gcs_job_info_stub.GetAllJobInfo(request, timeout=5)
```

Comment:
this is how `snapshot_head` gets driver info. Let me know if there's a better way.

Reply:
It looks like this is how it's done for legacy jobs, but for submission jobs we use a different API `self._job_info_client.get_all_jobs().items()` which ultimately pulls from the internal KV.  So would the alternative way to unify them be to store driver info in the internal KV and use `job_info_client` API?  

Anyways, it seems like your approach is simpler, and is async which is another advantage (though like Sang said, we should probably convert the `job_info_client` and `job_manager` APIs to use the async internal_kv API if they're going to be used in the API server.)

Reply:
I think this form is okay. We should just use state APIs after we migrate all these in the future